source/ConditionSet.cpp

```diff
@@ -325,477 +361,372 @@ set<string> ConditionSet::RelevantConditions() const
+	}
 
+	if(node.IsNumber(tokenNr))
+	{
+		expressionOperator = OP_LIT;
+		literal = node.Value(tokenNr);
+		++tokenNr;
+	}
+	else if(DataNode::IsConditionName(node.Token(tokenNr)))
+	{
+		expressionOperator = OP_VAR;
+		conditionName = node.Token(tokenNr);
+		++tokenNr;
+	}
+	else if(node.Token(tokenNr) == "(")
+	{
+		// We must already have handled an open-bracket to get here; this one goes into a sub-expression.
+		children.emplace_back();
+		children.back().ParseMini(node, tokenNr);
+	}
+	else
+		return FailParse(node, "expected terminal or open-bracket");
 
-// Convert the tokens and operators back to a string, for use in logging.
-const string ConditionSet::Expression::SubExpression::ToString() const
-{
-	string out;
-	static const string SPACE = " ";
-	size_t i = 0;
-	for( ; i < operators.size(); ++i)
+	// Keep parsing until we get to the closing bracket, if we had an open bracket.
+	while(hadOpenBracket)
 	{
-		if(!tokens[i].empty())
+		if(tokenNr >= node.Size())
+			return FailParse(node, "missing closing bracket");
+		else if(node.Token(tokenNr) == ")")
 		{
-			out += tokens[i];
-			out += SPACE;
+			// Remove the closing bracket.
+			++tokenNr;
+			hadOpenBracket = false;
+			// Make sure that this bracketed section gets used as a single terminal.
+			if(!PushDownFull(node))
+				return FailParse();
 		}
-		out += operators[i];
-		if(i != operators.size() - 1)
-			out += SPACE;
-	}
-	// The tokens vector contains more values than the operators vector.
-	for( ; i < tokens.size(); ++i)
-	{
-		if(i != 0)
-			out += SPACE;
-		out += tokens[i];
+		else
+			// If there are more tokens, then we need to have an infix operator here.
+			// Use the precedence of the AND operator, since we want to parse to the closing bracket.
+			if(!ParseFromInfix(node, tokenNr, OP_AND))
+				return FailParse();
 	}
-	return out;
+	return true;
 }
 
 
 
-// Interleave the tokens and operators, for use in the save file.
-const vector<string> ConditionSet::Expression::SubExpression::ToStrings() const
+bool ConditionSet::ParseFromInfix(const DataNode &node, int &tokenNr, ExpressionOp parentOp)
 {
-	auto out = vector<string>();
-	out.reserve(tokens.size() + operators.size());
-	size_t i = 0;
-	for( ; i < operators.size(); ++i)
+	// Keep on parsing until we reach an end-state (error, end-of-tokens, closing-bracket, lower precedence token)
+	while(true)
 	{
-		if(!tokens[i].empty())
-			out.emplace_back(tokens[i]);
-		out.emplace_back(operators[i]);
-	}
-	for( ; i < tokens.size(); ++i)
-		if(!tokens[i].empty())
-			out.emplace_back(tokens[i]);
-	return out;
-}
+		// At this point, we can expect one of the following:
+		// - an infix-operator
+		// - a closing bracket (hopefully matching an earlier open bracket)
+		// - end of the tokens.
+
+		// Reaching the end is fine, since we should have parsed a full terminal before this one.
+		// Reaching a closing bracket also means we are done (the parent should handle it).
```

Comment:
If I'm reading this correctly, the logic is:
 - get parsed by ParseMini, which takes the first opening bracket and passes the rest of the tokens to ParseFromInfix
 - which parses the token inside the brackets, and returns before the closing bracket
 - ParseMini sees the closing bracket and finishes.

The question is, what happens with nested brackets? I can't find any handling of those here.

Reply:
Directly nested brackets are handled on line 563, by ParseMini recursively calling another ParseMini.

Opening brackets after an operator are handled by ParseInfix recursively calling ParseMini to handle whatever there is after the operator.